test-harness/tests/test_cache.py

```diff
@@ -668,3 +668,249 @@ def test_if_early_boot_stage_can_recover_from_a_bit_slow_backend(
+                "state": "TASK_RUNNING",
+                }],
+            }]
+
+        self.assert_filter_regexp_for_invalid_apps(
+            filter_regexp, apps, nginx_class, mocker, superuser_user_header)
+
+    def test_app_without_task_ports(
+            self, nginx_class, mocker, superuser_user_header):
+        filter_regexp = {
+            "Cannot find ports for app '/without_ports'":
+                SearchCriteria(1, True),
+        }
+        apps = [{
+            "id": "/without_ports",
+            "labels": {
+                "DCOS_SERVICE_NAME": "without_ports",
+                "DCOS_SERVICE_SCHEME": "http",
+                "DCOS_SERVICE_PORT_INDEX": "0",
+                },
+            "tasks": [{
+                "id": "fake-id",
+                "state": "TASK_RUNNING",
+                "host": "10.0.0.1",
+                }],
+            }]
+
+        self.assert_filter_regexp_for_invalid_apps(
+            filter_regexp, apps, nginx_class, mocker, superuser_user_header)
+
+    def test_app_without_task_specified_port_idx(
+            self, nginx_class, mocker, superuser_user_header):
+        filter_regexp = {
+            "Cannot find port at Marathon port index '5' for app '/without_ports'":
+                SearchCriteria(1, True),
+        }
+        apps = [{
+            "id": "/without_ports",
+            "labels": {
+                "DCOS_SERVICE_NAME": "without_ports",
+                "DCOS_SERVICE_SCHEME": "http",
+                "DCOS_SERVICE_PORT_INDEX": "5",
+                },
+            "tasks": [{
+                "id": "fake-id",
+                "state": "TASK_RUNNING",
+                "host": "10.0.0.1",
+                "ports": [30000],
+                }],
+            }]
+
+        self.assert_filter_regexp_for_invalid_apps(
+            filter_regexp, apps, nginx_class, mocker, superuser_user_header)
+
+    def assert_filter_regexp_for_invalid_apps(
+            self,
+            filter_regexp,
+            apps,
+            nginx_class,
+            mocker,
+            auth_headers,
+            ):
+        """Helper method that will assert if provided regexp filter is found
+        in nginx logs for given apps response from Marathon upstream endpoint.
+
+        Arguments:
+            filter_regexp (dict): Filter definition where key is the message
+                looked up in logs and value is SearchCriteria definition
+            apps (list): List of apps that upstream endpoint should respond with
+            nginx_class (Nginx): NGINX fixture
+            mocker (Mocker): Mocker fixture
+            auth_header (dict): Headers that should be passed to NGINX request
+        """
+        ar = nginx_class()
+
+        mocker.send_command(endpoint_id='http://127.0.0.1:8080',
+                            func_name='apps_response',
+                            aux_data={"apps": apps})
+
+        url = ar.make_url_from_path('/service/test/foo/bar/')
```

Comment:
Lets be consistent here and use `/nginx-alwaysthere` task. It makes no difference - true, but it may make easier for other people to wrap their heads around test-harness.